File: native/android/app/src/main/java/com/smartmaps/ae/core/RoutingEngine.kt
package com.smartmaps.ae.core

import java.util.PriorityQueue
import kotlin.math.abs

/**
 * A/E Routing Engine — A* over the road graph + ETA + turn-by-turn.
 *
 * A* with a haversine heuristic over edge travel times (length / speed
 * limit), then instruction generation from bearing deltas at each node.
 * Everything is computed from graph geometry at request time — no
 * precomputed manoeuvre tables.
 */
class RoutingEngine {

    /** Route from [from] to [to] snapped onto [graph]. Null when unreachable. */
    fun route(graph: RoadGraph, from: GeoPoint, to: GeoPoint): Route? {
        if (graph.edges.isEmpty()) return null
        val startNode = nearestNode(graph, from) ?: return null
        val goalNode = nearestNode(graph, to) ?: return null
        if (startNode.id == goalNode.id) {
            return Route(listOf(from, to), arriveOnly(to), haversineM(from, to), 0.0)
        }

        // ── A* over travel-time cost ───────────────────────────────────────
        data class Entry(val nodeId: String, val f: Double)

        val gScore = mutableMapOf(startNode.id to 0.0)
        val cameFrom = mutableMapOf<String, Pair<String, RoadEdge>>() // node -> (prevNode, viaEdge)
        val open = PriorityQueue<Entry>(compareBy { it.f })
        open.add(Entry(startNode.id, heuristicS(startNode.point, goalNode.point)))
        val closed = mutableSetOf<String>()

        while (open.isNotEmpty()) {
            val current = open.poll()!!
            if (current.nodeId == goalNode.id) break
            if (!closed.add(current.nodeId)) continue

            val currentPoint = graph.nodes[current.nodeId]?.point ?: continue
            for (edge in graph.edgesFrom(current.nodeId)) {
                val neighborId = if (edge.fromNodeId == current.nodeId) edge.toNodeId else edge.fromNodeId
                if (neighborId in closed) continue
                val travelS = edge.lengthM / edge.speedLimitMps.coerceAtLeast(1.0)
                val tentative = (gScore[current.nodeId] ?: Double.MAX_VALUE) + travelS
                if (tentative < (gScore[neighborId] ?: Double.MAX_VALUE)) {
                    gScore[neighborId] = tentative
                    cameFrom[neighborId] = current.nodeId to edge
                    val neighborPoint = graph.nodes[neighborId]?.point ?: currentPoint
                    open.add(Entry(neighborId, tentative + heuristicS(neighborPoint, goalNode.point)))
                }
            }
        }
        if (goalNode.id !in gScore) return null

        // ── reconstruct node/edge chain ────────────────────────────────────
        val edgeChain = mutableListOf<RoadEdge>()
        val nodeChain = mutableListOf(goalNode.id)
        var cursor = goalNode.id
        while (cursor != startNode.id) {
            val (prev, via) = cameFrom[cursor] ?: return null
            edgeChain.add(0, via)
            nodeChain.add(0, prev)
            cursor = prev
        }

        // ── geometry: concatenate edge polylines in travel direction ───────
        val points = mutableListOf(from)
        for ((i, edge) in edgeChain.withIndex()) {
            val forward = edge.fromNodeId == nodeChain[i]
            val path = if (forward) edge.path else edge.path.reversed()
            points.addAll(path)
        }
        points.add(to)

        // ── turn-by-turn from bearing deltas at nodes ──────────────────────
        val steps = mutableListOf<RouteStep>()
        var cumDistM = haversineM(from, graph.nodes[startNode.id]!!.point)
        var cumEtaS = 0.0
        steps.add(RouteStep(Turn.DEPART, "Head ${cardinal(firstBearing(points))}", from, 0.0, 0.0))
        for (i in 1 until edgeChain.size) {
            val prevEdge = edgeChain[i - 1]
            val nextEdge = edgeChain[i]
            val nodePoint = graph.nodes[nodeChain[i]]?.point ?: continue
            cumDistM += prevEdge.lengthM
            cumEtaS += prevEdge.lengthM / prevEdge.speedLimitMps.coerceAtLeast(1.0)
            val inBearing = incomingBearing(prevEdge, nodeChain[i])
            val outBearing = outgoingBearing(nextEdge, nodeChain[i])
            val turn = classifyTurn(shortestAngleDeltaDeg(inBearing, outBearing))
            if (turn != Turn.CONTINUE) {
                val road = nextEdge.name ?: "the road"
                steps.add(RouteStep(turn, "${turnPhrase(turn)} onto $road", nodePoint, cumDistM, cumEtaS))
            }
        }
        val last = edgeChain.lastOrNull()
        if (last != null) {
            cumDistM += last.lengthM
            cumEtaS += last.lengthM / last.speedLimitMps.coerceAtLeast(1.0)
        }
        cumDistM += haversineM(graph.nodes[goalNode.id]!!.point, to)
        steps.add(RouteStep(Turn.ARRIVE, "Arrive at destination", to, cumDistM, cumEtaS))

        return Route(points, steps, cumDistM, cumEtaS)
    }

    /** Live remaining distance/ETA from a position on the route. */
    fun remaining(route: Route, at: GeoPoint): Pair<Double, Double> {
        var nearestIdx = 0
        var nearestD = Double.MAX_VALUE
        for (i in route.points.indices) {
            val d = haversineM(at, route.points[i])
            if (d < nearestD) {
                nearestD = d
                nearestIdx = i
            }
        }
        var dist = 0.0
        for (i in nearestIdx until route.points.size - 1) {
            dist += haversineM(route.points[i], route.points[i + 1])
        }
        val fraction = if (route.distanceM > 0) dist / route.distanceM else 0.0
        return dist to route.etaS * fraction
    }

    /** The next upcoming step from a position. */
    fun nextStep(route: Route, at: GeoPoint): RouteStep? {
        val (remainingM, _) = remaining(route, at)
        val travelledM = route.distanceM - remainingM
        return route.steps.firstOrNull { it.distanceM > travelledM + 5 } ?: route.steps.lastOrNull()
    }

    // ─── helpers ──────────────────────────────────────────────────────────────

    private fun heuristicS(a: GeoPoint, b: GeoPoint): Double = haversineM(a, b) / 33.0 // ~120 km/h ceiling keeps it admissible

    private fun nearestNode(graph: RoadGraph, p: GeoPoint): RoadNode? =
        graph.nodes.values.minByOrNull { haversineM(it.point, p) }

    private fun incomingBearing(edge: RoadEdge, atNodeId: String): Double {
        val path = if (edge.toNodeId == atNodeId) edge.path else edge.path.reversed()
        val n = path.size
        return bearingDeg(path[n - 2], path[n - 1])
    }

    private fun outgoingBearing(edge: RoadEdge, atNodeId: String): Double {
        val path = if (edge.fromNodeId == atNodeId) edge.path else edge.path.reversed()
        return bearingDeg(path[0], path[1])
    }

    private fun classifyTurn(delta: Double): Turn = when {
        abs(delta) < 20 -> Turn.CONTINUE
        delta in 20.0..45.0 -> Turn.SLIGHT_RIGHT
        delta in 45.0..120.0 -> Turn.RIGHT
        delta in 120.0..170.0 -> Turn.SHARP_RIGHT
        delta in -45.0..-20.0 -> Turn.SLIGHT_LEFT
        delta in -120.0..-45.0 -> Turn.LEFT
        delta in -170.0..-120.0 -> Turn.SHARP_LEFT
        else -> Turn.U_TURN
    }

    private fun turnPhrase(turn: Turn): String = when (turn) {
        Turn.SLIGHT_LEFT -> "Bear left"
        Turn.LEFT -> "Turn left"
        Turn.SHARP_LEFT -> "Turn sharp left"
        Turn.SLIGHT_RIGHT -> "Bear right"
        Turn.RIGHT -> "Turn right"
        Turn.SHARP_RIGHT -> "Turn sharp right"
        Turn.U_TURN -> "Make a U-turn"
        else -> "Continue"
    }

    private fun firstBearing(points: List<GeoPoint>): Double =
        if (points.size >= 2) bearingDeg(points[0], points[1]) else 0.0

    private fun cardinal(bearing: Double): String {
        val dirs = listOf("north", "north-east", "east", "south-east", "south", "south-west", "west", "north-west")
        return dirs[(((bearing + 22.5) % 360) / 45).toInt()]
    }

    private fun arriveOnly(to: GeoPoint) =
        listOf(RouteStep(Turn.ARRIVE, "Arrive at destination", to, 0.0, 0.0))
}

File: native/ios/SMCore/Core/RoutingEngine.swift
// RoutingEngine.swift — A/E Routing: A* + ETA + turn-by-turn.
//
// A* with a haversine heuristic over edge travel times, then instruction
// generation from bearing deltas at each node. Mirrors the Kotlin engine.

import Foundation

final class RoutingEngine {

    init() {}

    func route(graph: RoadGraph, from: GeoPoint, to: GeoPoint) -> Route? {
        guard !graph.edges.isEmpty,
              let startNode = nearestNode(graph, from),
              let goalNode = nearestNode(graph, to) else { return nil }
        if startNode.id == goalNode.id {
            return Route(
                points: [from, to],
                steps: [RouteStep(turn: .arrive, instruction: "Arrive at destination", point: to, distanceM: 0, etaS: 0)],
                distanceM: haversineM(from, to),
                etaS: 0
            )
        }

        // ── A* over travel-time cost ───────────────────────────────────────
        var gScore: [String: Double] = [startNode.id: 0]
        var cameFrom: [String: (prev: String, via: RoadEdge)] = [:]
        var open: [(nodeId: String, f: Double)] = [(startNode.id, heuristicS(startNode.point, goalNode.point))]
        var closed = Set<String>()

        while !open.isEmpty {
            open.sort { $0.f < $1.f }
            let current = open.removeFirst()
            if current.nodeId == goalNode.id { break }
            if closed.contains(current.nodeId) { continue }
            closed.insert(current.nodeId)

            guard let currentPoint = graph.nodes[current.nodeId]?.point else { continue }
            for edge in graph.edgesFrom(current.nodeId) {
                let neighborId = edge.fromNodeId == current.nodeId ? edge.toNodeId : edge.fromNodeId
                if closed.contains(neighborId) { continue }
                let travelS = edge.lengthM / max(edge.speedLimitMps, 1)
                let tentative = (gScore[current.nodeId] ?? .greatestFiniteMagnitude) + travelS
                if tentative < (gScore[neighborId] ?? .greatestFiniteMagnitude) {
                    gScore[neighborId] = tentative
                    cameFrom[neighborId] = (current.nodeId, edge)
                    let neighborPoint = graph.nodes[neighborId]?.point ?? currentPoint
                    open.append((neighborId, tentative + heuristicS(neighborPoint, goalNode.point)))
                }
            }
        }
        guard gScore[goalNode.id] != nil else { return nil }

        // ── reconstruct chain ──────────────────────────────────────────────
        var edgeChain: [RoadEdge] = []
        var nodeChain: [String] = [goalNode.id]
        var cursor = goalNode.id
        while cursor != startNode.id {
            guard let step = cameFrom[cursor] else { return nil }
            edgeChain.insert(step.via, at: 0)
            nodeChain.insert(step.prev, at: 0)
            cursor = step.prev
        }

        // ── geometry ───────────────────────────────────────────────────────
        var points: [GeoPoint] = [from]
        for (i, edge) in edgeChain.enumerated() {
            let forward = edge.fromNodeId == nodeChain[i]
            points.append(contentsOf: forward ? edge.path : edge.path.reversed())
        }
        points.append(to)

        // ── turn-by-turn ───────────────────────────────────────────────────
        var steps: [RouteStep] = []
        var cumDistM = haversineM(from, startNode.point)
        var cumEtaS = 0.0
        steps.append(RouteStep(
            turn: .depart,
            instruction: "Head \(cardinal(firstBearing(points)))",
            point: from, distanceM: 0, etaS: 0
        ))
        if edgeChain.count > 1 {
            for i in 1..<edgeChain.count {
                let prevEdge = edgeChain[i - 1]
                let nextEdge = edgeChain[i]
                guard let nodePoint = graph.nodes[nodeChain[i]]?.point else { continue }
                cumDistM += prevEdge.lengthM
                cumEtaS += prevEdge.lengthM / max(prevEdge.speedLimitMps, 1)
                let inB = incomingBearing(prevEdge, atNodeId: nodeChain[i])
                let outB = outgoingBearing(nextEdge, atNodeId: nodeChain[i])
                let turn = classifyTurn(shortestAngleDeltaDeg(from: inB, to: outB))
                if turn != .continue {
                    let road = nextEdge.name ?? "the road"
                    steps.append(RouteStep(
                        turn: turn,
                        instruction: "\(turnPhrase(turn)) onto \(road)",
                        point: nodePoint, distanceM: cumDistM, etaS: cumEtaS
                    ))
                }
            }
        }
        if let last = edgeChain.last {
            cumDistM += last.lengthM
            cumEtaS += last.lengthM / max(last.speedLimitMps, 1)
        }
        cumDistM += haversineM(goalNode.point, to)
        steps.append(RouteStep(turn: .arrive, instruction: "Arrive at destination", point: to, distanceM: cumDistM, etaS: cumEtaS))

        return Route(points: points, steps: steps, distanceM: cumDistM, etaS: cumEtaS)
    }

    /// Live remaining distance/ETA from a position on the route.
    func remaining(route: Route, at: GeoPoint) -> (distM: Double, etaS: Double) {
        var nearestIdx = 0
        var nearestD = Double.greatestFiniteMagnitude
        for (i, p) in route.points.enumerated() {
            let d = haversineM(at, p)
            if d < nearestD { nearestD = d; nearestIdx = i }
        }
        var dist = 0.0
        if nearestIdx < route.points.count - 1 {
            for i in nearestIdx..<(route.points.count - 1) {
                dist += haversineM(route.points[i], route.points[i + 1])
            }
        }
        let fraction = route.distanceM > 0 ? dist / route.distanceM : 0
        return (dist, route.etaS * fraction)
    }

    /// The next upcoming step from a position.
    func nextStep(route: Route, at: GeoPoint) -> RouteStep? {
        let travelledM = route.distanceM - remaining(route: route, at: at).distM
        return route.steps.first { $0.distanceM > travelledM + 5 } ?? route.steps.last
    }

    // ─── helpers ──────────────────────────────────────────────────────────────

    private func heuristicS(_ a: GeoPoint, _ b: GeoPoint) -> Double { haversineM(a, b) / 33.0 }

    private func nearestNode(_ graph: RoadGraph, _ p: GeoPoint) -> RoadNode? {
        graph.nodes.values.min { haversineM($0.point, p) < haversineM($1.point, p) }
    }

    private func incomingBearing(_ edge: RoadEdge, atNodeId: String) -> Double {
        let path = edge.toNodeId == atNodeId ? edge.path : edge.path.reversed()
        return bearingDeg(path[path.count - 2], path[path.count - 1])
    }

    private func outgoingBearing(_ edge: RoadEdge, atNodeId: String) -> Double {
        let path = edge.fromNodeId == atNodeId ? edge.path : edge.path.reversed()
        return bearingDeg(path[0], path[1])
    }

    private func classifyTurn(_ delta: Double) -> Turn {
        switch delta {
        case let d where abs(d) < 20: return .continue
        case 20...45: return .slightRight
        case 45...120: return .right
        case 120...170: return .sharpRight
        case -45...(-20): return .slightLeft
        case -120...(-45): return .left
        case -170...(-120): return .sharpLeft
        default: return .uTurn
        }
    }

    private func turnPhrase(_ turn: Turn) -> String {
        switch turn {
        case .slightLeft: return "Bear left"
        case .left: return "Turn left"
        case .sharpLeft: return "Turn sharp left"
        case .slightRight: return "Bear right"
        case .right: return "Turn right"
        case .sharpRight: return "Turn sharp right"
        case .uTurn: return "Make a U-turn"
        default: return "Continue"
        }
    }

    private func firstBearing(_ points: [GeoPoint]) -> Double {
        points.count >= 2 ? bearingDeg(points[0], points[1]) : 0
    }

    private func cardinal(_ bearing: Double) -> String {
        let dirs = ["north", "north-east", "east", "south-east", "south", "south-west", "west", "north-west"]
        return dirs[Int(((bearing + 22.5).truncatingRemainder(dividingBy: 360)) / 45)]
    }
}
